### PR TITLE
Modify sanitize_open() to use locked_file()

### PR DESCRIPTION
**Pull request by jakeogh**
Modify sanitize_open() to use locked_file(), preventing file corruption when two yt-dlp processes are attempting to write to the same file, where sys.platform != 'win32'.

With this patch, executing a second yt-dlp instance still results in a message: "[download] Resuming download at byte 2159224", however now the the next message is: "ERROR: unable to open for writing: [Errno 11] Resource temporarily unavailable". This fixes the current condition where the two processes would write to the same file, corrupting it. Now the second yt-dlp process moves on to the next required file, so in the case "Downloading 1 format(s): 303+251" if "*.f303.webm" is advisory locked by process 1, process 2 will move on to "*.f251.webm".

If a .part file is being used (--no-part is not enabled), there is a race condition if the .part file is closed by one process before it is renamed to remove '.part'. In that time window, the other process could modify the .part file before it's renamed. Using --no-part prevents this. It should be possible to lock the .part files.

There is no race condition where process 2 finishes downloading the last needed file (the audio, "*.f251.webm" in this example) after skipping the advisory locked video file ("*.f303.webm") that process 1 is writing to, because process 2 does not attempt to merge the two files together, instead, when process 1 completes it sees the already downloaded audio (*.f251.webm) and merges the two complete files together (usually) into a webm (because it [process 1] encountered no errors).

In the less common 'opposite' case, where process 1 (downloading the video 303) completes while process 2 is still writing the audio 251 file (and therefore has it advisory locked), the 303 process (process 1) throws the message "[download] Resuming download at byte 523264" and then exits with "ERROR: unable to open for writing: [Errno 11] Resource temporarily unavailable" because it is obeying the advisory lock process 2 has on "*.f251.webm". Unfortunately when the 251/audio process (process 2) completes, it does not attempt to merge the now complete video and audio files, since it has no way to know if the 303 file is complete (and it [process 2] already encountered an error trying to write to the 303 video file). To correct this, process 2 could check if another process still holds an advisory lock on the 303 file, and if not, obtain a lock on it and then compare the 303 file stat() with the expected metadata. This corner case is still left to fix. Re-executing a third yt-dlp process to the same target with the completed audio and video files (still) results in the expected merge of the two.

Untested on Windows, locking is not implemented, should be a unchanged if the included function signature modification is correct.

On other platforms there should be no case that results in a corrupt output file.

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

### CODE REVIEW

**Comment by pukkandan** [comment]
tick the appropriate fields in the template, and fix the linter issue

As I understand, the whole premise is to use two instances to download a single video so that each can download the separate formats and then one or the other merges them together. This is not a use case that I want to support/encourage. That said, locking the file is good either way, so this PR is good.

Are you able to implement this for windows as well? Having differing behavior b/w the OSes is not good

**Comment by jakeogh** [comment]
Thanks, boxes checked, linter fixed (I think). in this case my motivation is to deal with the accidental case where the user executes two instances inadvertently. Win32 is locking is outside my skill set, I don't use it. These changes do reflect testing done by a win32 user that tried the older (2018) version of this patch I attempted to merge to the original youtube-dl. It would be great if locking would also work on win32, but the subsystem is different, and as far as I can tell lock_file() is not able to accomplish this yet. I'm willing to attempt to fix if a win32 user can provide feedback.

File: yt_dlp/utils.py [hunk 1]
```diff
@@ -3462,31 +3464,36 @@ def _unlock_file(f):
     try:
         import fcntl
 
-        def _lock_file(f, exclusive):
-            fcntl.flock(f, fcntl.LOCK_EX if exclusive else fcntl.LOCK_SH)
+        def _lock_file(f, exclusive, block):
+            if block:
+                fcntl.flock(f, fcntl.LOCK_EX if exclusive else fcntl.LOCK_SH)
+            else:
+                fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB if exclusive else fcntl.LOCK_SH)
```
Comment (pukkandan) [inline yt_dlp/utils.py:3471]:
```suggestion
            fcntl.flock(f,
                        fcntl.LOCK_SH if not exclusive
                        else fcntl.LOCK_EX if block
                        else fcntl.LOCK_EX | fcntl.LOCK_NB)
```

File: yt_dlp/utils.py [hunk 2]
```diff
@@ -3439,7 +3441,7 @@ class OVERLAPPED(ctypes.Structure):
     whole_low = 0xffffffff
     whole_high = 0x7fffffff
 
-    def _lock_file(f, exclusive):
+    def _lock_file(f, exclusive, block=None):  # todo: block unused on win32
```
Comment (pukkandan) [inline yt_dlp/utils.py:3444]: keep the signatures exactly same. So no `=None`

File: yt_dlp/utils.py [hunk 3]
```diff
@@ -2075,7 +2076,8 @@ def sanitize_open(filename, open_mode):
             raise
         else:
             # An exception here should be caught in the caller
-            stream = open(encodeFilename(alt_filename), open_mode)
+            stream = locked_file(encodeFilename(filename), open_mode, block=False)
+            stream = stream.__enter__()
```
Comment (pukkandan) [inline yt_dlp/utils.py:2080]: bad design, but it's not your fault. The codebase should have used context managers

**Review by pukkandan — COMMENTED** [review]
For windows, I'll test it and let you know

File: yt_dlp/utils.py (hunk 2 above)
Comment (jakeogh) [reply yt_dlp/utils.py]: ah, yes, fixed.

**Comment by pukkandan** [comment]
please avoid force-pushing. it makes reviewing harder

**Comment by glenn-slayden** [comment]
fwiw, I already have a Win32 overhaul of locked_file in a local clone branch. It’s part of a full-blown ` --download-archive` rewrite that uses a persistent `set()` which is only reloaded when the disk file actually changes. A gigantic speedup for playlists and batch URLs.  Any interest?

**Comment by pukkandan** [comment]
> It’s part of a full-blown --download-archive rewrite that uses a persistent set() which is only reloaded when the disk file actually changes.

yt-dlp, unlike youtube-dl, already only loads the archive once. But we don't reload it if/when the file changes. That could be a nice improvement (separate from this PR)

> I already have a Win32 overhaul of locked_file in a local clone branch

Is it possible to isolate that code to be used in this PR?

**Comment by glenn-slayden** [comment]
> yt-dlp, unlike youtube-dl, already only loads the archive once. But we don't reload it if/when the file changes.

Looks like that change just finally went live earlier this month; I hadn’t picked it up yet.  I think my `set()` code would probably be simpler and possibly faster than the ArchiveTree, but it’s my own fault for missing out, due to being a total git n00b.

> Is it possible to isolate that code to be used in this PR

I’ll see if I can push a branch somehow for you to pillage.

**Comment by pukkandan** [comment]
> Looks like that change just finally went live earlier this month; I hadn’t picked it up yet

No, it was implemented in youtube-dlc itself 486ad2cd50d32482e33c370c89e7d40ebed3528d c71700dbe48891a5e3362c7034d77e7952874af3

**Comment by pukkandan** [comment]
@glenn-slayden 

> fwiw, I already have a Win32 overhaul of locked_file in a local clone branch. 
> I’ll see if I can push a branch somehow for you to pillage.

any update on this. I was keeping further review of this PR on hold so that any improvements you made can be added to it

**Comment by glenn-slayden** [comment]
Yes. Here is the code for the customized `os_lock/os_unlock` functions that I've had great success with for a while.

https://github.com/glenn-slayden/yt-dlp/blob/29dfb1e4509c47c6e0a8abfa98a8feb78d244f96/yt_dlp/utils.py#L3403-L3419

Turns out my changes were not that far from some Win32 interop code that was already in the ytdl family. However, the existing Python `locked_file` class was a bit too specific (or heavy-handed maybe?) for my usage. For one trivial example, there is an `assert` that forbids `b` mode which I wanted to use so I wouldn't be wasting time decoding the `bytes` to `str` under lock. Of course I could have just gotten rid of the assert, but there were a few other minor nits as well. Anyway I ended up just directly firing the operation in-situ from the call-site as shown below. 

I seem to recall that the `os_lock/os_unlock` functions have some special Win32 flags set up for a certain exclusive access mode whereby lock/unlock is only needed when *writing*. That is, it's not MRSW.

Example usage:

````
if not os.path.exists(os.path.dirname(a_file)):
    os.makedirs(os.path.dirname(a_file))
try:
    fileno = os.open(a_file, os.O_WRONLY|os.O_APPEND|os.O_BINARY|os.O_NOINHERIT)
    os_lock(fileno, 2)
    os.write(fileno, kb)
    os_unlock(fileno)
    os.fsync(fileno)
    os.close(fileno)
except IOError as ioe:
    raise Exception('uh-oh.')

````

As far as testing under actual file-exclusion conditions, the code was battered pretty heavily with decent contention over maybe two years of use without a single glitch. When active, this could be up to 6 `ytdl` instances continuously batching for long hours or days on the same machine, all of them always hitting the same `--download-archive` file (450,000 entries, 9.3 MB).

Now as you caused me to discover a couple days ago, I was quite shocked and flabbergasted to realize that when I started making the switch to **yt-dlp** earlier this year, I somehow neglected to port this `--download-archive` upgrade from my **youtube-dl** branch. Sure enough my archive rewrite has *not* been running, which has me a little worried that the code I've been running for months is not inherently unsuitable for 6-way contention. But that's my problem...

**Comment by glenn-slayden** [comment]
As far as porting over my `--download-archive` code, I'd very much like to, but I'm worried about making fairly intrusive changes when the `--download-archive` feature in dlp (unlike the situation was with ytdl) is not so desparate. 

Having said that, I judge that my implementation has two benefits over what you have now:

1. As I mentioned the other day, i believe hashing via `set()` will be faster than tree operations which are log(n) best-case;
2. Change detection in the contended file. Since this amounts to a correctness issue, it is by far the most serious reason to pursue this upgrade.

**Comment by jakeogh** [comment]
Win32 _lock_file() and _unlock_file() are defined in utils.py, this patch applies to git latest (it works fine here on linux) and as long as the current win32 locking code works, the behavior between *nix/win should be the same.

I was incorrect in my initial comment, *.part files are correctly locked (--no-part is not necessary, there is no race.)

The "opposite" (rare) case still exists, but generating a error is always better than the current situation where both processes silently write to the same file, corrupting it.

Note two yt-dlp processes, downloading different playlists can encounter this silent corruption of both happen to land on a video common to both sets.

**Comment by pukkandan** [comment]
> and locking the output file before downloading the *.part files. This would also fix the "opposite" case discussed below by locking the final merged A/V output file

No, lock must be applied ONLY to the files that are being currently modified

**Comment by jakeogh** [comment]
That is currently the case with this patch. I'm not following why taking an advisory lock on the final output file is a problem, it solves the corner cases, but it's not part of this patch anyway.

File: yt_dlp/utils.py [hunk 4]
```diff
@@ -3507,6 +3514,12 @@ def write(self, *args):
     def read(self, *args):
         return self.f.read(*args)
 
+    def flush(self, *args):
```
Comment (pukkandan) [inline yt_dlp/utils.py:3517]: Why *args?

File: yt_dlp/utils.py [hunk 5]
```diff
@@ -3462,31 +3464,36 @@ def _unlock_file(f):
...
-        assert mode in ['r', 'a', 'w']
+    def __init__(self, filename, mode, block=True, encoding=None):
+        assert mode in ['r', 'rb', 'a', 'ab', 'w', 'wb']
         self.f = io.open(filename, mode, encoding=encoding)
         self.mode = mode
+        self.block = block
 
     def __enter__(self):
-        exclusive = self.mode != 'r'
+        exclusive = self.mode not in ['r', 'rb']
```
Comment (pukkandan) [inline yt_dlp/utils.py:3494]:
```suggestion
        exclusive = 'r' not in self.mode
```

File: yt_dlp/utils.py [hunk 6]
```diff
@@ -2063,7 +2063,8 @@ def sanitize_open(filename, open_mode):
                 import msvcrt
                 msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)
             return (sys.stdout.buffer if hasattr(sys.stdout, 'buffer') else sys.stdout, filename)
-        stream = open(encodeFilename(filename), open_mode)
+        stream = locked_file(encodeFilename(filename), open_mode, block=False)
+        stream = stream.__enter__()
```
Comment (pukkandan) [inline yt_dlp/utils.py:2067]:
```suggestion
        stream = locked_file(filename, open_mode, block=False).open()
```
Perhaps this is better syntax (similar to `.close`)

(`encodeFilename` is a no-op)

File: yt_dlp/utils.py (hunk 4 above)
Comment (jakeogh) [reply yt_dlp/utils.py]: fixed

File: yt_dlp/utils.py (hunk 5 above)
Comment (jakeogh) [reply yt_dlp/utils.py]: fixed!

File: yt_dlp/utils.py (hunk 6 above)
Comment (jakeogh) [reply yt_dlp/utils.py]: Change made, as a future TODO `class locked_file()` is a context manager, and as you mentioned https://github.com/yt-dlp/yt-dlp/pull/1066#discussion_r714647420 it's not being used as one.

File: yt_dlp/utils.py [hunk 7]
```diff
@@ -2183,6 +2188,15 @@ def write(self, *args):
...
         return self.f.read(*args)
 
+    def flush(self):
+        self.f.flush()
+
+    def open(self):
+        return self.__enter__()
+
+    def close(self, *args):
+        self.__exit__(self, *args, value=False, traceback=False)
```
Comment (pukkandan) [inline yt_dlp/utils.py:2198]:
```suggestion
    def close(self):
        self.__exit__(self, value=False, traceback=False)
```
right?

**Comment by RedirectCleaner** [comment]
this commit introduced the following regression

https://github.com/yt-dlp/yt-dlp/issues/2670

**Pull request merged**